domain/src/main/java/org/oppia/domain/oppialogger/crashlytics/ExceptionsController.kt

```diff
@@ -0,0 +1,174 @@
+package org.oppia.domain.oppialogger.crashlytics
+
+import androidx.lifecycle.LiveData
+import org.oppia.app.model.ExceptionLog
+import org.oppia.app.model.OppiaExceptionLogs
+import org.oppia.data.persistence.PersistentCacheStore
+import org.oppia.domain.oppialogger.ExceptionLogStorageCacheSize
+import org.oppia.util.data.AsyncResult
+import org.oppia.util.data.DataProviders
+import org.oppia.util.logging.ConsoleLogger
+import org.oppia.util.logging.ExceptionLogger
+import org.oppia.util.networking.NetworkConnectionUtil
+import org.oppia.util.system.OppiaClock
+import javax.inject.Inject
+
+class ExceptionsController @Inject constructor(
+  private val exceptionLogger: ExceptionLogger,
+  private val dataProviders: DataProviders,
+  cacheStoreFactory: PersistentCacheStore.Factory,
+  private val consoleLogger: ConsoleLogger,
+  private val networkConnectionUtil: NetworkConnectionUtil,
+  private val oppiaClock: OppiaClock,
+  @ExceptionLogStorageCacheSize private val exceptionLogStorageCacheSize: Int
+) {
+  private val exceptionLogStore =
+    cacheStoreFactory.create("exception_logs", OppiaExceptionLogs.getDefaultInstance())
+
+  lateinit var exceptionCause: Throwable
+
+  fun logException(exception: Exception, timestamp: Long) {
+    uploadOrCacheExceptionLog(exception, timestamp)
+  }
+
+  /**
+   * Checks network connectivity of the device.
+   *
+   * Saves the [exceptionLog] to the [exceptionLogStore] in the absence of it.
+   * Uploads to remote service in the presence of it.
+   */
+  private fun uploadOrCacheExceptionLog(exception: Exception, timestamp: Long) {
+    when (networkConnectionUtil.getCurrentConnectionStatus()) {
+      NetworkConnectionUtil.ConnectionStatus.NONE ->
+        cacheExceptionLog(exceptionToExceptionLog(exception, timestamp))
+      else -> exceptionLogger.logException(exception)
+    }
+  }
+
+  /** Returns an [ExceptionLog] from an [exception]. */
+  private fun exceptionToExceptionLog(exception: Exception, timestamp: Long): ExceptionLog {
+    val exceptionLogBuilder = ExceptionLog.newBuilder()
+    val exceptionLogCauseBuilder = ExceptionLog.newBuilder()
```

Comment:
Why not recurse a call to the same function, instead? E.g.

```kotlin
exception.cause?.let {
  exceptionLogBuilder.cause = exceptionToExceptionLog(it, timestamp)
}
```

It seems like that would reduce a lot of the construction that's needed here.

Reply:
Yep, done.